src/coprocessor_v2/endpoint.rs

```diff
@@ -26,18 +34,74 @@ impl<E: Engine> Endpoint<E> {
+        req: coprv2pb::RawCoprocessorRequest,
     ) -> impl Future<Output = coprv2pb::RawCoprocessorResponse> {
-        todo!("Coprocessor V2 is currently not implemented.");
+        let mut response = coprv2pb::RawCoprocessorResponse::default();
+
+        let coprocessor_result = self.handle_request_impl(storage, req);
 
-        // Make sure we produce a valid return type
-        // (because `todo!()` doesn't work with `impl Trait`).
-        #[allow(unreachable_code)]
-        std::future::ready(coprv2pb::RawCoprocessorResponse::default())
+        match coprocessor_result {
+            Ok(data) => response.set_data(data),
+            Err(CoprocessorError::RegionError(region_err)) => response.set_region_error(region_err),
+            Err(CoprocessorError::Other(o)) => response.set_other_error(o),
+        }
+
+        std::future::ready(response)
+    }
+
+    #[inline]
+    fn handle_request_impl<E: Engine, L: LockManager>(
+        &self,
+        storage: &Storage<E, L>,
+        req: coprv2pb::RawCoprocessorRequest,
+    ) -> Result<RawResponse, CoprocessorError> {
+        let plugin =
+            self.plugin_registry
+                .get_plugin(&req.copr_name)
+                .ok_or(CoprocessorError::Other(format!(
+                    "No registered coprocessor with name '{}'",
+                    req.copr_name
+                )))?;
+
+        let raw_storage_api = RawStorageImpl::new(req.get_context(), storage);
+        let region = Region {
+            id: req.get_context().get_region_id(),
+            start_key: vec![], // TODO how to get start_key and end_key?
+            end_key: vec![],   // TODO
+            region_epoch: RegionEpoch {
+                conf_ver: req.get_context().get_region_epoch().get_conf_ver(),
+                version: req.get_context().get_region_epoch().get_version(),
+            },
+        };
+
+        let plugin_result = plugin.on_raw_coprocessor_request(&region, &req.data, &raw_storage_api);
+
+        plugin_result.map_err(|err| {
+            if let Some(region_err) = extract_region_error(&err) {
+                CoprocessorError::RegionError(region_err)
+            } else {
+                CoprocessorError::Other(format!("{}", err))
+            }
+        })
     }
 }
 
-#[cfg(test)]
-mod tests {}
+fn extract_region_error(error: &PluginError) -> Option<kvproto::errorpb::Error> {
+    match error {
+        PluginError::StorageError(storage_err) => {
+            match storage_err {
+                coprocessor_plugin_api::StorageError::Other(other_err) => other_err
+                    .downcast_ref::<storage::Result<()>>()
+                    .map_or(None, |e| storage::errors::extract_region_error::<()>(e)),
+                // TODO: KeyNotInRegion is not set in https://tikv.github.io/doc/src/tikv/storage/errors.rs.html#193-233
+                // so I assume we should also not set it?
+                coprocessor_plugin_api::StorageError::KeyNotInRegion { .. } => None,
```

Comment:
I'm not entirely sure which errors we should pass back to the client as proper errors and which ones as "error messages".

There is this function `storage::errors::extract_region_error()` that is used to set specific errors from TiKV when returning the result to the client and other errors are just converted to an error message. 
For "normal" `TiKVClient::raw_put()`, ... requests, `KeyNotInRegion` is converted to a string and only the error message is returned. I do the same here, but I was not sure about this.